File: src/DAO/BackgroundsDAO.py
import json
import datetime
from typing import Any
from src.DAO.MiscClasses import (
    SkillProficiencies,
    LanguageProficiencies,
    StartingEquipment,
    Source,
    ToolProficiencies,
    AdditionalSpells,
    Prerequisite,
    SkillToolLanguageProficiencies,
)


class Background:
    """
    Has a _copy
    """

    name: str
    source: str
    page: int
    srd: bool
    basic_rules: bool
    skill_proficiencies: SkillProficiencies
    language_proficiencies: LanguageProficiencies
    starting_equipment: StartingEquipment
    entries: list
    has_fluff: bool
    tool_proficiencies: ToolProficiencies
    feats: list[str]
    from_feature: list[str]
    has_fluff_images: bool
    additional_spells: AdditionalSpells
    additional_sources: list[Source]
    prerequisite: Prerequisite
    skill_tool_language_proficiencies: SkillToolLanguageProficiencies
    other_sources: list[Source]

    def __init__(self, object: dict[str, Any]) -> None:
        self.name = object["name"] if "name" in object.keys() else ""
        self.source = object["source"] if "source" in object.keys() else ""
        self.page = object["page"] if "page" in object.keys() else -1
        self.srd = object["srd"] if "srd" in object.keys() else False
        self.basicRules = object["basicRules"] if "basicRules" in object.keys() else ""
        self.skill_proficiencies = (
            SkillProficiencies(object["skillProficiencies"][0])
            if "skillProficiencies" in object.keys()
            else SkillProficiencies(None)
        )
        self.language_proficiencies = (
            LanguageProficiencies(object["languageProficiencies"][0])
            if "languageProficiencies" in object.keys()
            else LanguageProficiencies(None)
        )
        self.starting_equipment = (
            StartingEquipment(object["startingEquipment"])
            if "startingEquipment" in object.keys()
            else StartingEquipment([])
        )
        self.entries = object["entries"] if "entries" in object.keys() else []
        self.has_fluff = object["hasFluff"] if "hasFluff" in object.keys() else False
        self.tool_proficiencies = (
            ToolProficiencies(object["toolProficiencies"])
            if "toolProficiencies" in object.keys()
            else ToolProficiencies(None)
        )
        self.feats = []
        if 'feats' in object.keys():
            for feat_obj in object['feats']:
                self.feats.extend([str(key) for key in feat_obj.keys()])
        self.from_feature = [str(key) for key in object["fromFeature"].keys()] if "fromFeature" in object.keys() else []
        self.has_fluff_images = object["hasFluffImages"] if "hasFluffImages" in object.keys() else False
        self.additional_spells = (
            AdditionalSpells(object["additionalSpells"])
            if "additionalSpells" in object.keys()
            else AdditionalSpells(None)
        )
        self.additional_sources = (
            [Source(source) for source in object["additionalSources"]] if "additionalSources" in object.keys() else []
        )
        self.prerequisite = (
            Prerequisite(object["prerequisite"]) if "prerequisite" in object.keys() else Prerequisite(None)
        )
        self.skill_tool_language_proficiencies = (
            SkillToolLanguageProficiencies(object["skillToolLanguageProficiencies"])
            if "skillToolLanguageProficiencies" in object.keys()
            else SkillToolLanguageProficiencies(None)
        )
        self.other_sources = (
            [Source(source) for source in object["otherSources"]] if "otherSources" in object.keys() else []
        )

    def as_dict(self) -> dict[str, Any]:
        """This is effectivaly read only as it creates a copy

        Returns:
            dict[str, Any]: the class as a dict
        """

        copy = self.__dict__.copy()
        copy["skill_proficiencies"] = self.skill_proficiencies.__dict__
        copy["language_proficiencies"] = self.language_proficiencies.__dict__
        copy["starting_equipment"] = self.starting_equipment.as_dict()
        copy["tool_proficiencies"] = self.tool_proficiencies.__dict__
        copy["additional_spells"] = self.additional_spells.__dict__
        copy["additional_sources"] = [source.__dict__ for source in self.additional_sources]
        copy["prerequisite"] = self.prerequisite.as_dict()
        copy["skill_tool_language_proficiencies"] = self.skill_tool_language_proficiencies.as_dict()
        copy["other_sources"] = [source.__dict__ for source in self.other_sources]
        return copy


class Backgrounds:
    backgrounds: dict[str, Background]

    def __init__(self) -> None:
        with open("data/raw/backgrounds.json") as f:
            obj = json.load(f)

            self.backgrounds = {}
            for background in obj["background"]:
                self.backgrounds[background["name"]] = Background(background)

    def as_dict(self) -> dict[str, Any]:
        """This is effectivaly read only as it creates a copy

        Returns:
            dict[str, Any]: the class as a dict
        """

        copy = self.__dict__.copy()
        for background_name in self.backgrounds.keys():
            copy["backgrounds"][background_name] = self.backgrounds[background_name].as_dict()
        return copy


if __name__ == "__main__":
    # times = []
    # count = 1000
    # for _ in range(0, count):
    #     start = datetime.datetime.now()
    #     backgrounds = Backgrounds()
    #     end = datetime.datetime.now()
    #     times.append((end - start).microseconds)
    #     del backgrounds

    # average_time = sum(times) / count

    # print(f"average time was {average_time} microseconds")
    backgrounds = Backgrounds()
    print(json.dumps(backgrounds.as_dict(), indent=4))

File: src/dao/backgrounds_dao.rs
use serde_json::{Map, Value};

use super::common::{
    form_key, serde_as_array, serde_as_array_mapping, serde_as_bool, serde_as_object, serde_as_object_from_option, serde_as_string, AdditionalSpells, Entry, LanguageProficiencies, Prerequisite, SkillProficiencies, SkillToolLanguageProficiencies, Source, StartingEquipment, ToolProficiencies,
};

/**
 * Has a _copy
 */
#[derive(Debug)]
pub struct Background {
    pub name: String,
    pub source: Source,
    pub key: String,
    pub srd: bool,
    pub basic_rules: bool,
    pub skill_proficiencies: SkillProficiencies,
    pub language_proficiencies: LanguageProficiencies,
    pub starting_equipment: StartingEquipment,
    pub entries: Vec<Entry>,
    pub has_fluff: bool,
    pub tool_proficiencies: ToolProficiencies,
    pub feats: Vec<String>,
    pub from_feature: Vec<String>,
    pub has_fluff_images: bool,
    pub additional_spells: AdditionalSpells,
    pub additional_sources: Vec<Source>,
    pub prerequisite: Prerequisite,
    pub skill_tool_language_proficiencies: SkillToolLanguageProficiencies,
    pub other_sources: Vec<Source>,
}

impl Background {
    pub fn new(object: Value) -> Background {
        let p_object = serde_as_object(&object, Map::new());

        let name = serde_as_string(p_object.get("name"), "N/A".to_string());
        let source = Source::new(p_object.get("source"), p_object.get("page"));

        return Background {
            key: form_key(&name, &source.name),
            name,
            source,
            srd: serde_as_bool(p_object.get("srd"), false),
            basic_rules: serde_as_bool(p_object.get("basicRules"), false),
            skill_proficiencies: SkillProficiencies::new(p_object.get("skillProficiencies")),
            language_proficiencies: LanguageProficiencies::new(p_object.get("languageProficiencies")),
            starting_equipment: StartingEquipment::new(p_object.get("startingEquipment")),
            entries: serde_as_array(p_object.get("entries")).iter().map(|i| Entry::new(i)).collect(),
            has_fluff: serde_as_bool(p_object.get("hasFluff"), false),
            tool_proficiencies: ToolProficiencies::new(p_object.get("toolProficiencies")),
            feats: serde_as_array_mapping(p_object.get("feats"), serde_as_string, "N/A".to_string()),
            from_feature: serde_as_array_mapping(p_object.get("fromFeature"), serde_as_string, "N/A".to_string()),
            has_fluff_images: serde_as_bool(p_object.get("hasFluffImages"), false),
            additional_spells: AdditionalSpells::new(p_object.get("additionalSpells")),
            additional_sources: serde_as_array_mapping(p_object.get("additionalSources"), serde_as_object_from_option, Map::new()).iter().map(|i| Source::new(i.get("source"), i.get("page"))).collect(),
            prerequisite: Prerequisite::new(p_object.get("prerequisite")),
            skill_tool_language_proficiencies: SkillToolLanguageProficiencies::new(p_object.get("skillToolLanguageProficiencies")),
            other_sources: serde_as_array_mapping(p_object.get("otherSources"), serde_as_object_from_option, Map::new()).iter().map(|i| Source::new(i.get("source"), i.get("page"))).collect(),
        };
    }
}
